prov/shm/src/smr_progress.c

```diff
@@ -381,9 +376,17 @@ static int smr_progress_cmd_msg(struct smr_ep *ep, struct smr_cmd *cmd)
 			"unidentified operation type\n");
 		err = -FI_EINVAL;
 	}
-	ret = smr_complete_rx(ep, entry->context, cmd->msg.hdr.op,
-			cmd->msg.hdr.op_flags | (entry->flags & ~SMR_MULTI_RECV),
-			total_len, entry->iov[0].iov_base, cmd->msg.hdr.addr,
+
+	comp_entry = *entry;
+	if (entry->flags & SMR_MULTI_RECV) {
+		smr_progress_multi_recv(ep, recv_queue, entry, &comp_entry,
+					total_len);
+		free_entry = (comp_entry.flags == entry->flags);
+	}
+
+	ret = smr_complete_rx(ep, comp_entry.context, cmd->msg.hdr.op,
+			cmd->msg.hdr.op_flags | comp_entry.flags,
+			total_len, comp_entry.iov[0].iov_base, cmd->msg.hdr.addr,
```

Comment:
The iov for the duplicate is the same as the original entry, I think.  The duplicate ep_entry doesn't seem necessary, since all this code seems to care about is whether some flag has been set or not.

The setting of free_entry = entry->flags == duplicate entry->flags is indirect.  The entry is being freed if it is not a multi-recv msg, or if it is but the space left is smaller than the minimal size.  A more explicit coding of this would be clearer, rather than hiding it behind a function and masked flags.